File: algorithms/medium/longest_consecutive_sequence.ex
defmodule Solution do
  @spec longest_consecutive(nums :: [integer]) :: integer
  def longest_consecutive(nums) do
    aux(nums, %{}, 0)
  end

  defp aux([], _map, ans), do: ans

  defp aux([x | xs], map, ans) do
    if Map.has_key?(map, x) do
      aux(xs, map, ans)
    else
      left = Map.get(map, x - 1, 0)
      right = Map.get(map, x + 1, 0)
      len = left + right + 1
      new_map = map |> Map.put(x, len) |> Map.put(x - left, len) |> Map.put(x + right, len)
      aux(xs, new_map, max(ans, len))
    end
  end
end

File: algorithms/medium/longest_consecutive_sequence.rs
use std::collections::HashMap;

impl Solution {
    pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
        let mut map = HashMap::new();
        let mut ans = 0;
        for &n in &nums {
            if map.contains_key(&n) {
                continue;
            }
            let left = *map.get(&(n - 1)).unwrap_or(&0);
            let right = *map.get(&(n + 1)).unwrap_or(&0);
            let curr = left + right + 1;
            map.insert(n, curr);
            ans = ans.max(curr);
            map.insert(n - left, curr);
            map.insert(n + right, curr);
        }
        ans
    }
}
